### DISCUSSION TITLE
How to export and import a database inside supabase

### DISCUSSION THREAD

**Discussion by SandroMaglione** [supabase/supabase#773] [Questions] [answered]
Hi,

I would like to export a database I created in one project in supabase (scheme+data) and then import it into another project still in supabase.
I saw #666 and #175, but it is not clear to me where should I write the commands (e.g. `pg_restore`) in supabase, or even if it would be the correct approach.

How can I achieve that? Is it possible?

Great work by the way, the project is amazing 👍

**Comment by steve-chavez** [comment, marked as answer]
You have the database connection string on `/settings/database`, you copy it and then you can connect to your db from your terminal(`psql <connection_string>`) or any other client.

Then:

 `pg_dump <connection_string_from_db_1>` -> extract the dump ->  `pg_restore -d  <connection_string_from_db_2> dump`

The databases share some tables/schemas, so a full dump could fail. An option would be to `pg_dump` specific tables that you want and only restore them on the other database.

**Reply by SandroMaglione** [reply]
Thanks @steve-chavez 

I ended up downloading [PostgreSQL](https://www.postgresql.org/download/). Then I used **pgAdmin** to connect to the server using the `host` URL from `/settings/database` from my account supabase, and the password of the project. Then I simply made a [backup+restore of the database](https://stackoverflow.com/a/48780919/7033357).

**Reply by SandroMaglione** [reply]
I also used [this](https://stackoverflow.com/a/13823560/7033357) to remove the older database and restore the newest one.

**Reply by MentalGear** [reply]
Are custom functions/triggers restored as well that way?

**Reply by sagar-jani** [reply]
It seems like pg_restore needs the dump in compressed binary format otherwise it gives this error - pg_restore: error: input file appears to be a text format dump. Please use psql.

**Reply by babblebey** [reply]
I was able to get this done using the import/export option after connecting to supabase db via pgadmin as stated here [Connecting with pgAdmin](https://supabase.com/docs/guides/database/pgadmin) as my requirement was to just export a particular table only

**Comment by jdcauley** [comment]
Extending this convo: Attempted a schema dump in command line and in pgAdmin based on the docs and then this discussion and got the following errors:

```
pg_dump: error: query failed: ERROR:  permission denied for table identities
pg_dump: error: query was: LOCK TABLE auth.identities IN ACCESS SHARE MODE
```

![](http://avngr.us/NlbS5K+)

**Reply by kangmingtay** [reply]
Hi @jdcauley, can you please email `support@supabase.io` with your project ref so we can take a look at the permissions being set? Thanks!

**Reply by awalias** [reply]
can you try this from the SQL editor in the dashboard:
```
GRANT ALL ON TABLE auth.identities TO postgres;
```

**Reply by jdcauley** [reply]
@awalias I replied via email, I got a permission denied error.

**Reply by gillkyle** [reply]
In order to get this to work I had to keep appending `GRANT` statements in the SQL editor. What finally worked was the following: 

```
GRANT ALL ON TABLE auth.identities TO postgres;
GRANT ALL ON TABLE realtime.schema_migrations TO postgres;
GRANT ALL ON TABLE realtime.subscription TO postgres;
GRANT ALL ON SEQUENCE refresh_tokens_id_seq TO postgres;
GRANT ALL ON SEQUENCE realtime.subscription_id_seq TO postgres;
```

And then `pg_dump <connection_string_from_db> > backup.sql`

**Reply by riderx** [reply]
Even with that, i couldn't backup i had to add this:
```
GRANT ALL ON TABLE storage.migrations TO postgres;
GRANT ALL ON TABLE supabase_functions.migrations TO postgres;
GRANT ALL ON TABLE supabase_functions.hooks TO postgres;
GRANT USAGE, SELECT ON ALL SEQUENCES IN SCHEMA public TO postgres;
GRANT USAGE, SELECT ON ALL SEQUENCES IN SCHEMA supabase_functions TO postgres;
GRANT USAGE, SELECT ON ALL SEQUENCES IN SCHEMA storage TO postgres;
GRANT USAGE, SELECT ON ALL SEQUENCES IN SCHEMA auth TO postgres;
```
And now export work but i can't restore i get :
```
Restore database 'postgres2'
Restoring...
restore_pg_13.0: while PROCESSING TOC:
restore_pg_13.0: from TOC entry 12; 2615 16468 SCHEMA auth supabase_admin
restore_pg_13.0: 
error: could not execute query: ERROR:  must be member of role "supabase_admin"
Command was: ALTER SCHEMA auth OWNER TO supabase_admin;
```
in new table and in existing table:
```
Restore database 'postgres'
Restoring...
restore_pg_13.0: while PROCESSING TOC:
restore_pg_13.0: from TOC entry 12; 2615 16468 SCHEMA auth supabase_admin
restore_pg_13.0: error: could not execute query: ERROR:  schema "auth" already exists
Command was: CREATE SCHEMA auth;
```

**Reply by riderx** [reply]
So i copied table by table it was looking good, just missing the RLS, but then sequence were broken and even i found a way to fix it now real-time feature fail to work.

**Comment by simonh1000** [comment]
I need to do this too. I know very little about SQL so what I did was go to Supabase SQL Editor on the web. Then, in the top section I wrote `pg_dump  /settings/database` (based on the accepted answer) and clicked `run`. I got `Failed to validate sql query: syntax error at or near "pg_dump"`.  I have no doubt that that is completely wrong, but I don't really know where to begin

**Reply by logemann** [reply]
pg_dump is a CLI utility shipped with postgres, you cant run this as an SQL. You can install pg_dump on your machine together with pg_restore and then work with those commands.

**Comment by sunilguptasg** [comment]
the /settings/database is the menu path - you go there in your supabase admin console and copy the info...

**Comment by LucasKuhn** [comment]
Writing here in case it helps somebody else - I did a backup from a Heroku postgres and wanted to move it to Supabase. What worked for me was using PgAdmin.
1. Create Supabase project
2. Download PgAdmin, right click on servers -> Register -> Server 
3. Open your Supabase project -> Project Settings -> Database 
4. On PgAdmin, paste the Host and Password
5. Open your new server, right click on postgres database -> Restore
6. Select your sql dump 
7. On Data/Object, do not save Owner and Privilege
8. Click Restore 

<img width="575" alt="image" src="https://user-images.githubusercontent.com/26127185/236242383-838e8197-57b2-4eba-89d4-609bef0fa854.png">

**Comment by revskill10** [comment]
Why not adding a button in console to export json schema from a table (with child table) ?

**Reply by Snailedlt** [reply]
Would be very useful to have something like what pgAdmin has, where you can extract databases or tables to varying file formats.

in pgAdmin you can backup your db/table like this:
![image](https://github.com/supabase/supabase/assets/43886029/c4a37ce7-8afc-4a80-8e3c-eec4118fef23)
![image](https://github.com/supabase/supabase/assets/43886029/692dab1c-08ef-46b4-9c15-2b95996b5024)

**Comment by simonh1000** [comment]
Can't help thinking this is a completely obvious thing for supabase users to want to do, and that there should be an out of the box way of making such a copy. Perhaps some would not be permitted by data protection rules to make copy of data but, for most, the ability to work safely with a copy of production data is useful. Perhaps the assumption is that we are all postgres cli experts? But there's a javascript SDK which means the audience is broader than postgres folk

**Reply by didier** [reply]
Need something like this!

**Comment by GorvGoyl** [comment]
Was able to backup and restore like this:

SETUP
- Install [postgres](https://www.postgresql.org/download/) and then add it to system path

For mac:

```sh
brew install postgresql@17
# add to path:
# .zshrc
export PATH="/opt/homebrew/opt/postgresql@17/bin:$PATH"
```

BACKUP (schema + data):

- Get connection string: Supabase project settings -> Connect -> URI with Session Pooler -> copy string
- note: if there's `@` symbol in db password then replace it with `%40`
```sql
pg_dump [connection_string_with_password] \
  --clean \
  --if-exists \
  --quote-all-identifiers \
  --no-owner \
  --disable-triggers \
  --no-privileges \
  > postgresdump.sql
```

RESTORE (schema + data): 
- Get connection string of restore location: Supabase project settings -> Connect -> URI with Session Pooler -> copy string
- There were some cli errors when running this command but data was still intact.

```sql
psql -d [connection_string_with_password] -f postgresdump.sql
```

**Reply by knandersen** [reply]
in case someone comes by because they get status 500 errors when trying to get storage items, I had the same issue after restoring my database. In my case, because it wasn't a terrible amount of files, I simply:

1. Downloaded all files from my restored bucket
2. Deleted the restored bucket
3. Created the restored bucket
4. Uploaded all files to the newly created bucket

No issues afterwards.

**Comment by vikito755** [comment]
Does the database dump preserve RLS policies or is there something extra Supabase does to achieve that?

**Reply by ElectricFern** [reply]
Looks like you have to manage and migrate those separate.
https://supabase.com/docs/guides/platform/migrating-within-supabase/backup-restore#schema-changes-to-auth-and-storage

**Reply by Klrfl** [reply]
If you already have RLS rules on the database itself, making a dump would preserve the rules, I assume. At least it does when you make migrations via the CLI

**Comment by highway11** [comment]
I'm struggling to get a working backup and restore from either of these backup and restore methods.

1. Backup using the [script from this post](https://gist.github.com/ethicnology/4aa30be9f93daf7c38bdd43d8c281ea3) https://github.com/orgs/supabase/discussions/22200:  
  Restore using:  using: `psql -h [HOST] -p 5432 -d postgres -U postgres.your-tenant-id < public.sql`

The above restore errors out with many errors, the main one being:` ERROR:  must be member of role "supabase_admin"` I don't get a working restore

I have also tried this backup method:
```
pg_dump -h 127.0.0.1 -U postgres.your-tenant-id -p 5432 -d postgres  \
  --clean \
  --if-exists \
  --quote-all-identifiers \
  --no-owner \
  --disable-triggers \
  --no-privileges \
  > postgresdump.sql
```
paired with this restore:
`psql -h 127.0.0.1 -U postgres.your-tenant-id -p 5432 -d postgres -f postgresdump.sql`
However this method also throws many errors and I get a broken app, where I can no longer login.

Has anyone found a reliable method for both backing up and restoring an entire supabase database, including auth, storage, public, functions, triggers, basically everything in the database? This needs to work on self hosted. Do I need to drop the database before attempting to restore from the 2nd backup?
